### ISSUE TITLE
Trouble with cunit, regex_t from pcreposix.h, and regcomp() from libc

### ISSUE THREAD

**Issue by dilyanpalauzov** [cyrusimap/cyrus-imapd#2891]
On my system
```c
#include <regex.h>
printf("%lu\n", sizeof(regex_t));
```
emits 64 and
```c
#include <pcreposix.h>
printf("%lu\n", sizeof(regex_t));
```
prints 24.

So in glob.c is crucial, when regcomp() is from libc, that regex.h is used also from libc. (If regcomp is from libc, but malloc(sizeof(regex_t)) is from pcre, then the allocated size is too small).

On my system cunit/unit is linked with:

```
/bin/bash ./libtool  --tag=CC   --mode=link clang  -O0 -g  -L/usr/local/lib -licuuc -licudata -fuse-ld=bfd -L/usr/local/lib -o cunit/unit cunit/unit.o cunit/syslog.o cunit/timeout.o cunit/timezones.o cunit/timeofday.o cunit/aaa-db.o cunit/annotate.o cunit/backend.o cunit/binhex.o cunit/bitvector.o cunit/buf.o cunit/byteorder64.o cunit/charset.o cunit/command.o cunit/conversations.o cunit/crc32.o cunit/dlist.o cunit/duplicate.o cunit/getxstring.o cunit/glob.o cunit/guid.o cunit/hash.o cunit/imapurl.o cunit/libconfig.o cunit/mboxname.o cunit/md5.o cunit/message.o cunit/msgid.o cunit/parseaddr.o cunit/parse.o cunit/prot.o cunit/ptrarray.o cunit/quota.o cunit/rfc822tok.o cunit/search_expr.o cunit/seqset.o cunit/sieve.o cunit/spool.o cunit/squat.o cunit/strarray.o cunit/strconcat.o cunit/times.o cunit/tok.o cunit/vparse.o imap/mutex_fake.o imap/spool.o sieve/libcyrus_sieve.la lib/libcyrus.la lib/libcyrus_min.la -L/usr/local/lib -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support    -lpcre -lpcreposix -lz -L/usr/local/lib -lxml2 -L/usr/local/lib -lical -licalss -licalvcal -L/usr/local/lib -licui18n -licuuc -licudata  -lsqlite3 -lsasl2 -L/usr/local/lib64 -lssl -lcrypto -lssl -lcrypto   imap/libcyrus_imap.la lib/libcyrus.la lib/libcyrus_min.la -L/usr/local/lib -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support    -lpcre -lpcreposix -lz -L/usr/local/lib -lxml2 -L/usr/local/lib -lical -licalss -licalvcal -L/usr/local/lib -licui18n -licuuc -licudata  -lsqlite3 -lsasl2 -L/usr/local/lib64 -lssl -lcrypto -lssl -lcrypto   com_err/et/libcyrus_com_err.la -lcunit -L/usr/local/lib -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support    -lpcre -lpcreposix -lz -L/usr/local/lib -lxml2 -L/usr/local/lib -lical -licalss -licalvcal -L/usr/local/lib -licui18n -licuuc -licudata  -lsqlite3

libtool: link: clang -O0 -g -fuse-ld=bfd -o cunit/unit cunit/unit.o cunit/syslog.o cunit/timeout.o cunit/timezones.o cunit/timeofday.o cunit/aaa-db.o cunit/annotate.o cunit/backend.o cunit/binhex.o cunit/bitvector.o cunit/buf.o cunit/byteorder64.o cunit/charset.o cunit/command.o cunit/conversations.o cunit/crc32.o cunit/dlist.o cunit/duplicate.o cunit/getxstring.o cunit/glob.o cunit/guid.o cunit/hash.o cunit/imapurl.o cunit/libconfig.o cunit/mboxname.o cunit/md5.o cunit/message.o cunit/msgid.o cunit/parseaddr.o cunit/parse.o cunit/prot.o cunit/ptrarray.o cunit/quota.o cunit/rfc822tok.o cunit/search_expr.o cunit/seqset.o cunit/sieve.o cunit/spool.o cunit/squat.o cunit/strarray.o cunit/strconcat.o cunit/times.o cunit/tok.o cunit/vparse.o imap/mutex_fake.o imap/spool.o  -L/usr/local/lib sieve/.libs/libcyrus_sieve.a -L/usr/local/lib64 imap/.libs/libcyrus_imap.a -ljansson /git/cyrus/cyrus-imapd.30/sieve/.libs/libcyrus_sieve.a /git/cyrus/cyrus-imapd.30/com_err/et/.libs/libcyrus_com_err.a /git/cyrus/cyrus-imapd.30/lib/.libs/libcyrus_min.a /git/cyrus/cyrus-imapd.30/lib/.libs/libcyrus.a /usr/local/lib/libxapian.so -lrt /usr/local/lib/../lib64/libstdc++.so lib/.libs/libcyrus.a lib/.libs/libcyrus_min.a /lib/libuuid.so /usr/local/lib/libsasl2.so -lresolv -lssl -lcrypto com_err/et/.libs/libcyrus_com_err.a /usr/local/lib/libcunit.so -L/src/CUnit-2.1-3/CUnit/Sources -lc -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support /usr/local/lib/libpcreposix.so /usr/local/lib/libpcre.so /usr/local/lib/libxml2.so /usr/local/lib/liblzma.so -lical -licalss -licalvcal -licui18n -licuuc -licudata /usr/local/lib/libsqlite3.so -lz -lm -ldl -lpthread -pthread -Wl,-rpath -Wl,/usr/local/lib/../lib64 -Wl,-rpath -Wl,/usr/local/lib/../lib64
```

readelf -d cunit/unit prints
```
 0x0000000000000001 (NEEDED)             Shared library: [libjansson.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libxapian.so.30]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libuuid.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libsasl2.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libresolv.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libssl.so.1.1]
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto.so.1.1]
 0x0000000000000001 (NEEDED)             Shared library: [libcunit.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgssapi_krb5.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libkrb5.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libk5crypto.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libcom_err.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libkrb5support.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpcreposix.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpcre.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libxml2.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [liblzma.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libical.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libicalss.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libicalvcal.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libicui18n.so.63]
 0x0000000000000001 (NEEDED)             Shared library: [libicuuc.so.63]
 0x0000000000000001 (NEEDED)             Shared library: [libicudata.so.63]
 0x0000000000000001 (NEEDED)             Shared library: [libsqlite3.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x000000000000000f (RPATH)              Library rpath: [/usr/local/lib/../lib64]
```
so libpcre is searched for symbols, only after libc has not offered the symbols.

When I run `LD_DEBUG=all cunit/unit annotate` it prints
```
     21466:	symbol=regcomp;  lookup in file=cunit/unit [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/libjansson.so.4 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/libxapian.so.30 [0]
     21466:	symbol=regcomp;  lookup in file=/lib64/librt.so.1 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/../lib64/libstdc++.so.6 [0]
     21466:	symbol=regcomp;  lookup in file=/lib/libuuid.so.1 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/libsasl2.so.3 [0]
     21466:	symbol=regcomp;  lookup in file=/lib64/libresolv.so.2 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/../lib64/libssl.so.1.1 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/../lib64/libcrypto.so.1.1 [0]
     21466:	symbol=regcomp;  lookup in file=/usr/local/lib/libcunit.so.1 [0]
     21466:	symbol=regcomp;  lookup in file=/lib64/libc.so.6 [0]
     21466:	binding file cunit/unit [0] to /lib64/libc.so.6 [0]: normal symbol `regcomp' [GLIBC_2.2.5]
```
so `regcomp()` is taken from libc (requires 64bytes for regex_t), but at compile time regex_t comes from libpcre (so is 24 bytes).

I do not know how to enforce reordering of the libraries, but with this order regcomp() from glob_init() works on 24 allocated bytes, but it needs 64

Passing --disable-pcre to ./configure does not define ENABLE_REGEX and nothing works. It would have worked, if --disable-pcre was not passed and PCRE was not installed.

**Comment by elliefm** [comment]
The PCRE library is known to be a huge nuisance.  Debian has been patching around it for 18+ years, and I think RedHat/Fedora and probably some others also apply Debian's patch.  The PCRE maintainers refuse to integrate the patch, so if you use glibc (or any other libc that integrates the POSIX regex functions directly) and PCRE on the same system and don't patch around it, you will have trouble.  See also #2629 and #1731, and there's also a few mailing list threads on the subject

We would like to rip out support for PCRE entirely and use some other library that doesn't conflict with glibc and refuse to change (RE2 has been suggested), but that hasn't happened yet.  Another possibility would be dropping support for POSIX regexes, and then only using the PCRE-specific interfaces (which do not clash with the POSIX ones in glibc).  Which would probably be nearly as much work, and also hasn't happened.

It would be good to fix the configure options so that you can disable pcre on a system with pcre installed, and still have POSIX regex functionality.  But I think each time this comes up, people just end up applying Debian's patch to PCRE instead.

**Comment by elliefm** [comment]
> Passing --disable-pcre to ./configure does not define ENABLE_REGEX and nothing works. It would have worked, if --disable-pcre was not passed and PCRE was not installed.

Looks like this was already fixed on master, so I've just cherry-picked the commits to 3.0 (74eb4d266 and f5a5d018f).  With those commits, you should be able to just `--disable-pcre` and get glibc's regexes only.  It should also detect and use rxposix if you have that.

**Comment by dilyanpalauzov** [comment]
Why isn’t the prce’s POSIX extended regex removed completely from cyrus, so that the implementation uses POSIX extended regex from libc?

**Comment by elliefm** [comment]
When configure.ac looks for pcre, it also checks specifically for utf8 support, so I guess that's the feature we really care about.  I guess glibc or some other systems' POSIX implementations didn't support utf8 at the time.  No idea if this is still the case?

I built 3.0 and master the other day, both with `--disable-pcre` and just using my system's (debian stretch) glibc regexes, and cassandane didn't detect any new problems that it didn't already find with pcre enabled.  But I don't know if this means that the glibc implementation is adequate these days, or if our cassandane tests are inadequate to find the specific issues we selected pcre to address in the first place...

**Comment by dilyanpalauzov** [comment]
Duplicate of  #2629 and #1731.

**Comment by dilyanpalauzov** [comment]
So the goal is to:
* support UTF-8, which excludes libc
* not use libpcreposix.so and libpcre2-posix.so, as their ABI collides with libc’s
* use RE2 or non-posix PCRE

**Comment by dilyanpalauzov** [comment]
Or regular expressions with ICU, this does not introduce new dependancy.
